### ISSUE TITLE
ssl certificate does not validate when using verify-full

### ISSUE THREAD

**Issue by flixman** [openssl/openssl#20101]
I have reported this issue in the project where I found it, psycopg2, but I was asked to report it here. In case this is misplaced... Do you have any hint on where should I report it? I am trying to connect to a postgresql:15.1-alpine docker container using the following parameters:

```python
ssl_options = dict(sslcert="TLS/client/client.crt", sslkey="TLS/client/private/client.key",
                   sslrootcert="TLS/ca/ca.crt", sslmode="verify-full", ssl_min_protocol_version="TLSv1.3")
admin_engine = create_engine('postgresql://{}:{}@{}/{}'.format(
    os.environ['POSTGRES_USER'], os.environ['POSTGRES_PASSWORD'], os.environ['POSTGRES_HOST'],
    os.environ['POSTGRES_DB']), connect_args=ssl_options, isolation_level='AUTOCOMMIT')
```
It fails with the following error:
```bash
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "127.0.0.1", port 5432 failed: server certificate for "localhost" does not match host name "127.0.0.1"
```

However: should I connect to the same server with this connection string, all works:
```bash
psql "port=5432 host=localhost user=xxxx sslcert=TLS/client/client.crt sslkey=TLS/client/private/client.key sslrootcert=TLS/ca/ca.crt sslmode=verify-full dbname=postgres"
```

If I switch the sslmode to `verify-ca`, then it works in both cases... but I think it should work also in the case of `verify-full`. With respect to the server certificate, these are the relevant sections (in case it matters, I want to highlight I am not setting any CN in the subject):

```bash
        Subject: C = ES, ST = Spain, L = Barcelona, O = MyCompany, OU = Default CA Deployment
        X509v3 extensions:
            X509v3 Authority Key Identifier: 
                8E:D4:DE:DD:5D:0D:22:D1:BA:3F:AB:F4:69:42:80:2F:B8:FC:6D:B2
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Key Usage: 
                Digital Signature, Non Repudiation, Key Encipherment, Data Encipherment
            X509v3 Subject Alternative Name: 
                IP Address:127.0.0.0, IP Address:127.0.0.1, DNS:localhost
            X509v3 Subject Key Identifier: 
                E4:94:13:93:50:60:F1:D2:EB:B9:EC:28:03:A9:C2:5E:CC:4F:0C:C9
```

These are the versions of the packages/OS/modules I am using:
- OS: Arch Linux
- Psycopg version: 2.9.5
- Python version: 3.10.9
- PostgreSQL version: 15.1
- pip version: 22.2.2
- openssl: 3.0.7

**Comment by t8m** [comment]
The message `connection to server at "127.0.0.1", port 5432 failed: server certificate for "localhost" does not match host name "127.0.0.1"` does not come from OpenSSL.

You can try to use `openssl s_client -connect 127.0.0.1:5432 -CAfile TLS/ca/ca.crt -cert TLS/ca/ca.crt -key -key TLS/client/private/client.key -starttls postgres` to see what openssl reports.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
